### DISCUSSION TITLE
AssemblyLoadContext Unloadability

### DISCUSSION THREAD

**Discussion by lukaspirkl** [AvaloniaUI/Avalonia#12286] [Q&A]
I would like to create an application with plugins that would be able to load and unload plugins on the fly. The plugin contains UserControl which is placed as a child into the Border control. I am able to load it without any problem but unloading is problematic, as [described in the documentation](https://learn.microsoft.com/en-us/dotnet/standard/assembly/unloadability). 

When I am not adding the plugin's UserControl into the UI the unload is working. So it looks like I am not holding any reference to block the unloading. I was trying to get some information from WinDbg but I have no idea what I am doing. It looks like there is a generated class `CompiledAvaloniaXaml.!AvaloniaResources+NamespaceInfo:/UserControl1.axaml` that may be the problem. The UserControl1 is the name of my UserControl inside the plugin.

Before I spend more time with that, I would like to ask whether this use case is supported or if I am boldly going where no one has gone before.

**Comment by timunie** [comment]
can you wire up a minimal sample for us to explore? then maybe (can't promise) someone will have an idea what is missing. Probably it's also a bug, but can't say yet.

**Reply by lukaspirkl** [reply]
Here it is: [DotNetPlugins.zip](https://github.com/AvaloniaUI/Avalonia/files/12143271/DotNetPlugins.zip)

It is not possible to unload the AssemblyLoadContext right after the UserControl from the plugin is instantiated. I guess the `AvaloniaXamlLoader.Load(this);` inside the `InitializeComponent` is caching the compiled xaml and therefore it is not possible to unload it.

**Reply by timunie** [reply]
Thx for the sample. @maxkatz6 may have more insights here.
